src/control/server/instance.go

```diff
@@ -230,21 +224,38 @@ func (srv *IOServerInstance) determineRank(ctx context.Context, ready *srvpb.Not
 		}
 	}
 
+	// If the join was processed locally, we don't need to
+	// perform more steps, so let the caller know.
+	if resp.LocalJoin {
+		r = system.NilRank
+	}
+
 	return r, nil
 }
 
-// setRank determines the instance rank and sends a SetRank dRPC request
+// joinSystem determines the instance rank and sends a SetRank dRPC request
 // to the IOServer.
-func (srv *IOServerInstance) setRank(ctx context.Context, ready *srvpb.NotifyReadyReq) error {
+func (srv *IOServerInstance) joinSystem(ctx context.Context, ready *srvpb.NotifyReadyReq) error {
 	r, err := srv.determineRank(ctx, ready)
 	if err != nil {
 		return err
 	}
 
+	// hack -- If we receive a nil rank from determineRank() it means
+	// that we can skip the rest of setup because it already happened.
+	if r.Equals(system.NilRank) {
```

Comment:
would prefer to just return Boolean from determineRank indicating a local join rather than assuming from a NilRank

Reply:
Yeah that's fair. When I added that I wasn't sure yet if this was going to be a short-term or long-term solution. I think I'm pretty happy with it as a long-term solution though so it's worth doing correctly.